Sprint 6 solution

MaximusDenisov-YP opened the pull request:
Закончил выполнение задания 6-го спринта.

Commit:
Начало выполнения задания спринта №6
- Реализовал вариант связного списка для хранения истории задач.
- Добавил в методы удаления задач из основного менеджера последовательное удаление их же из менеджера истории задач.
- Добавил класс Node, для работы реализации связного списка.
- Добавил /out в .gitignore
- Убрал лишний метод из интерфейса HistoryManager и его реализации соответственно.

Commit:
Закончил выполнения задания спринта №6
- Выполнил доп. задание.
- Исправил тест, исправил удаление subTasks при удалении Epic наполненный ими.

Commit:
Исправил замечание по проверке CheckStyle.
Не было пробела перед + в return toString для Epic.class

Commit:
Исправление замечания no such file in .test directory

find: ‘./test’: No such file or directory

Commit:
Повторное исправление замечания no such file in .test directory.

Попытка переместить папку с тестами на уровень src.

find: ‘./test’: No such file or directory

avfyodorov commented on src/main/ru/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -171,17 +171,24 @@ public Subtask updateSubtask(Subtask subTask) {
 
```
```java
public void deleteAllTasks() {
    public void deleteAllSubtasks() {
    public void deleteAllEpics() {
```

В этих методах также нужно не забыть почистить историю.

avfyodorov commented on src/main/ru/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,29 +2,73 @@
+    static class LinkedHistoryList {
+
+        Map<Integer, Node> historyMap = new HashMap<>();
+        public int size = 0;
```
public int size = 0;

Это поле точно не нужно, для того чтобы узнать размер истории можно воспользоваться методом 

historyMap.size()

avfyodorov commented on src/main/ru/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,29 +2,73 @@
+        historyList.removeNode(id);
+    }
+
+    static class LinkedHistoryList {
```
class LinkedHistoryList {

В принципе, без этого класса можно было бы и обойтись, а его методы сделать приватными методами менеджера истории. 

На Ваше усмотрение. 

avfyodorov commented on src/main/ru/kanban/manager/InMemoryHistoryManager.java:
```diff
@@ -2,29 +2,73 @@
+
+        ArrayList<Task> getTasks() {
+            ArrayList<Task> result = new ArrayList<>();
+            for (Node node : historyMap.values()) {
```
Лучше сформировать список, чтобы задачи располагались по порядку:
```java
	public List<Task> getHistory() {
		ArrayList<Task> tasks = new ArrayList<>();

		Node node = first;
		while (node != null) {
			tasks.add(node.task);
			node = node.next;
		}
		return tasks;
	}

```

avfyodorov commented on src/main/ru/kanban/manager/Node.java:
```diff
@@ -0,0 +1,21 @@
+
+import ru.kanban.entity.Task;
+
+public class Node {
```
Этот класс можно было бы сделать внутренним классом менеджера истории, потому что он всё равно больше нигде не нужен. 

На усмотрение.

avfyodorov commented on src/main/ru/kanban/manager/Node.java:
```diff
@@ -0,0 +1,21 @@
+    Node next;
+    Node prev;
+
+    Node(Node prev, Task element, Node next) {
```
Но в любом случае, даже если решите оставить, полям и методам нужно добавить модификаторы доступа.

avfyodorov reviewed (commented):
Добрый день, Максим! 

Хорошая работа, от меня всего пара уточнений.

Commit:
Исправил код по замечаниям от ревьюера.

- Ввиду изменений был улучшен вид кода в целом, сокращено кол-во строк кода и всё стало более понятным и читабельным.

- Node был переведён во вложенный класс InMemoryHistoryManager, т.к. более нигде применяться не будет.

- Добавлены модификаторы новым полям InMemoryHistoryManager.

- Попробовал использовать функциональный стиль для циклов в InMemoryTaskManager, тяжёлой логики внутри методов не наблюдается, потому выглядит более лаконично и менее шумно.

- Исправил логику удаления задач из истории задач, при методах удаления all.

avfyodorov reviewed (commented):
Добрый день, Максим!

Замечаний нет.
Работа принята.
